File: src/Runtime/PHP/Optimizers/Adam.php
<?php

namespace PHP2xAI\Runtime\PHP\Optimizers; 

use PHP2xAI\Runtime\PHP\Core\GraphRuntime;

class Adam extends Optimizer
{
	protected float $learningRate = 0.1;
	protected float $beta1 = 0.9;
	protected float $beta2 = 0.999;
	protected float $eps = 0.00000001;
	
	protected array $m = [];
	protected array $v = [];
	
	protected array $mp = []; // array of previous $m
	protected array $vp = []; // array of previuss $v
	
	protected $stepNumber = 1;
	
	public function __construct(float $learningRate = 0.1, float $beta1 = 0.9, float $beta2 = 0.999, float $eps = 0.00000001)
	{
		$this->learningRate = $learningRate;
		$this->beta1 = $beta1;
		$this->beta2 = $beta2;
		$this->eps = $eps;
	}
	
	public function getConfig() : array
	{
		return array(
			"name"		=>	"Adam",
			"params"	=>	array(
				"learningRate"	=>	$this->learningRate,
				"beta1"			=>	$this->beta1,
				"beta2"			=>	$this->beta2,
				"eps"			=>	$this->eps,
			),
		);
	}
	
	public function step(GraphRuntime $graph)
	{
		$n = max(1, $graph->accSteps);
		// echo $graph->accSteps."\n";
		$beta1PowT = pow($this->beta1, $this->stepNumber);
		$beta2PowT = pow($this->beta2, $this->stepNumber);
		
		foreach ($graph->trainable as $tid)
		{
			$t = $graph->tensors[$tid];
			
			$size = count($t->data);
			
			for ($i = 0; $i < $size; $i++)
			{
				$g = $t->grad[$i] / $n;
				
				if ($this->gradClip !== null)
				{
					$clip = $this->gradClip;
					if ($g > $clip) $g = $clip;
					else if ($g < -$clip) $g = -$clip;
				}
				
				$mtp = $this->mp[$tid][$i] ?? 0;
				$vtp = $this->vp[$tid][$i] ?? 0;
				
				$mt = $this->beta1 * $mtp + (1 - $this->beta1) * $g;
				$vt = $this->beta2 * $vtp + (1 - $this->beta2) * ($g * $g);
				
				$this->mp[$tid][$i] = $mt;
				$this->vp[$tid][$i] = $vt;
				
				$mt_ = $mt / (1 - $beta1PowT);
				$vt_ = $vt / (1 - $beta2PowT);
				
				$t->data[$i] -= $this->learningRate * ($mt_ / (sqrt($vt_) + $this->eps));
			}
		}
		
		$this->stepNumber++;
	}
}

File: src/Runtime/CPP/Optimizers/Adam.cpp
#include <algorithm>
#include <cmath>
#include <stdexcept>

#include "Adam.hpp"
#include "../Core/runtime.hpp"

namespace PHP2xAI::Runtime::CPP::Optimizers
{
	Adam::Adam(Scalar learningRate, Scalar beta1, Scalar beta2, Scalar eps)
		: learningRate_(learningRate),
		beta1_(beta1),
		beta2_(beta2),
		eps_(eps)
	{
	}

	void Adam::step(GraphRuntime& graph)
	{
		const auto beta1PowT = std::pow(beta1_, static_cast<Scalar>(stepNumber_));
		const auto beta2PowT = std::pow(beta2_, static_cast<Scalar>(stepNumber_));

		const std::vector<int> &trainable = graph.getTrainableTensorIds();
		for (std::size_t parameter = 0; parameter < trainable.size(); ++parameter)
		{
			const int tid = trainable[parameter];
			if (!graph.tensorHasFloatingPointDType(tid))
				throw std::runtime_error("Adam can only update floating point tensors");

			auto& mVec = mp_[tid];
			auto& vVec = vp_[tid];

			const std::size_t size = graph.getTensorSize(tid);
			if (mVec.size() < size)
				mVec.resize(size, static_cast<Scalar>(0));
			if (vVec.size() < size)
				vVec.resize(size, static_cast<Scalar>(0));

			for (std::size_t i = 0; i < size; ++i)
			{
				Scalar g = graph.getTensorGradValue(tid, i);

				if (gradClip_)
				{
					const auto clip = *gradClip_;
					if (g > clip)
						g = clip;
					else if (g < -clip)
						g = -clip;
				}

				const auto mtp = mVec[i];
				const auto vtp = vVec[i];

				const auto mt = beta1_ * mtp + (1 - beta1_) * g;
				const auto vt = beta2_ * vtp + (1 - beta2_) * (g * g);

				mVec[i] = mt;
				vVec[i] = vt;

				const auto mtHat = mt / (1 - beta1PowT);
				const auto vtHat = vt / (1 - beta2PowT);

				const Scalar value = graph.getTensorDataValue(tid, i);
				graph.setTensorDataValue(tid, i, value - learningRate_ * (mtHat / (std::sqrt(vtHat) + eps_)));
			}
		}

		++stepNumber_;
	}
}
